File: formal/runtime/java/dafny/DafnySet.java
// Copyright by the contributors to the Dafny Project
// SPDX-License-Identifier: MIT
//
// EACL target-cost refinement:
// The upstream Dafny Java runtime copies a java.util.HashSet for every
// immutable set union. EACL's generated indexed traversal grows sets one
// element at a time, making the otherwise linear traversal quadratic.
// This drop-in implementation preserves the DafnySet API and extensional
// semantics while using Clojure's persistent hash set. Clojure is already a
// mandatory EACL runtime dependency.

package dafny;

import clojure.lang.IPersistentSet;
import clojure.lang.PersistentHashSet;
import java.util.*;

public class DafnySet<T> {
    private IPersistentSet innerSet;

    public DafnySet() {
        innerSet = PersistentHashSet.EMPTY;
    }

    public DafnySet(Set<T> s) {
        this((Collection<T>) s);
    }

    public DafnySet(Collection<T> c) {
        assert c != null : "Precondition Violation";
        innerSet = PersistentHashSet.EMPTY;
        for (T value : c) {
            innerSet = (IPersistentSet) innerSet.cons(value);
        }
    }

    public DafnySet(DafnySet<T> other) {
        assert other != null : "Precondition Violation";
        innerSet = other.innerSet;
    }

    public DafnySet(List<T> l) {
        this((Collection<T>) l);
    }

    private DafnySet(IPersistentSet persistentSet, boolean trusted) {
        innerSet = persistentSet;
    }

    @SafeVarargs
    public static <T> DafnySet<T> of(T ... elements) {
        DafnySet<T> result = new DafnySet<>();
        for (T element : elements) {
            result.add(element);
        }
        return result;
    }

    private static final DafnySet<Object> EMPTY =
        new DafnySet<>(PersistentHashSet.EMPTY, true);

    @SuppressWarnings("unchecked")
    public static <T> DafnySet<T> empty() {
        return (DafnySet<T>) EMPTY;
    }

    @SuppressWarnings("unchecked")
    public static <T> TypeDescriptor<DafnySet<? extends T>> _typeDescriptor(
            TypeDescriptor<T> elementType) {
        return TypeDescriptor.referenceWithDefault(
                (Class<DafnySet<? extends T>>) (Class<?>) DafnySet.class,
                DafnySet.empty());
    }

    public boolean isSubsetOf(DafnySet other) {
        assert other != null : "Precondition Violation";
        return other.containsAll(this);
    }

    public boolean isProperSubsetOf(DafnySet other) {
        assert other != null : "Precondition Violation";
        return isSubsetOf(other) && size() < other.size();
    }

    public boolean contains(Object value) {
        assert value != null : "Precondition Violation";
        return innerSet.contains(value);
    }

    public <U> boolean disjoint(DafnySet<? extends U> other) {
        assert other != null : "Precondition Violation";
        DafnySet<?> smaller = size() <= other.size() ? this : other;
        DafnySet<?> larger = smaller == this ? other : this;
        for (Object value : smaller.Elements()) {
            if (larger.contains(value)) {
                return false;
            }
        }
        return true;
    }

    @SuppressWarnings("unchecked")
    public static <T> DafnySet<T> union(
            DafnySet<? extends T> left,
            DafnySet<? extends T> right) {
        assert left != null : "Precondition Violation";
        assert right != null : "Precondition Violation";
        if (left.isEmpty()) {
            return (DafnySet<T>) right;
        }
        if (right.isEmpty()) {
            return (DafnySet<T>) left;
        }
        DafnySet<? extends T> larger =
            left.size() >= right.size() ? left : right;
        DafnySet<? extends T> smaller = larger == left ? right : left;
        DafnySet<T> result =
            new DafnySet<>((IPersistentSet) larger.innerSet, true);
        for (T value : smaller.Elements()) {
            result.add(value);
        }
        return result;
    }

    @SuppressWarnings("unchecked")
    public static <T> DafnySet<T> difference(
            DafnySet<? extends T> left,
            DafnySet<? extends T> right) {
        assert left != null : "Precondition Violation";
        assert right != null : "Precondition Violation";
        if (left.isEmpty() || right.isEmpty()) {
            return (DafnySet<T>) left;
        }
        DafnySet<T> result =
            new DafnySet<>((IPersistentSet) left.innerSet, true);
        for (T value : right.Elements()) {
            result.remove(value);
        }
        return result;
    }

    public static <T> DafnySet<T> intersection(
            DafnySet<? extends T> left,
            DafnySet<? extends T> right) {
        assert left != null : "Precondition Violation";
        assert right != null : "Precondition Violation";
        DafnySet<? extends T> smaller =
            left.size() <= right.size() ? left : right;
        DafnySet<? extends T> larger = smaller == left ? right : left;
        DafnySet<T> result = new DafnySet<>();
        for (T value : smaller.Elements()) {
            if (larger.contains(value)) {
                result.add(value);
            }
        }
        return result;
    }

    public boolean containsAll(DafnySet other) {
        assert other != null : "Precondition Violation";
        for (Object value : other.Elements()) {
            if (!contains(value)) {
                return false;
            }
        }
        return true;
    }

    public int size() {
        return innerSet.count();
    }

    public int cardinalityInt() {
        return size();
    }

    public boolean isEmpty() {
        return size() == 0;
    }

    // The compiler uses mutating construction only on a fresh wrapper.
    // The underlying persistent value is never mutated.
    public boolean add(T value) {
        assert value != null : "Precondition Violation";
        boolean changed = !innerSet.contains(value);
        innerSet = (IPersistentSet) innerSet.cons(value);
        return changed;
    }

    public boolean remove(T value) {
        assert value != null : "Precondition Violation";
        boolean changed = innerSet.contains(value);
        innerSet = innerSet.disjoin(value);
        return changed;
    }

    public boolean removeAll(DafnySet<T> other) {
        assert other != null : "Precondition Violation";
        boolean changed = false;
        for (T value : other.Elements()) {
            changed |= remove(value);
        }
        return changed;
    }

    public boolean addAll(DafnySet<T> other) {
        assert other != null : "Precondition Violation";
        boolean changed = false;
        for (T value : other.Elements()) {
            changed |= add(value);
        }
        return changed;
    }

    public Collection<DafnySet<T>> AllSubsets() {
        List<T> elements = new ArrayList<>(Elements());
        int count = elements.size();
        HashSet<DafnySet<T>> result = new HashSet<>();
        for (int mask = 0; mask < (1 << count); mask++) {
            DafnySet<T> subset = new DafnySet<>();
            for (int index = 0; index < count; index++) {
                if ((mask & (1 << index)) != 0) {
                    subset.add(elements.get(index));
                }
            }
            result.add(subset);
        }
        return result;
    }

    @Override
    public boolean equals(Object value) {
        if (this == value) {
            return true;
        }
        if (value == null || getClass() != value.getClass()) {
            return false;
        }
        DafnySet<?> other = (DafnySet<?>) value;
        return size() == other.size() && containsAll(other);
    }

    @Override
    public int hashCode() {
        return innerSet.hashCode();
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder("{");
        String separator = "";
        for (T element : Elements()) {
            result.append(separator).append(Helpers.toString(element));
            separator = ", ";
        }
        return result.append("}").toString();
    }

    public DafnyMultiset<T> asDafnyMultiset() {
        return new DafnyMultiset<>(Elements());
    }

    @SuppressWarnings("unchecked")
    public Set<T> Elements() {
        return (Set<T>) innerSet;
    }
}

File: formal/runtime/js/dafny-set.js
  // EACL target-cost refinement:
  // Dafny 4.11 represents JavaScript sets as arrays. Membership and immutable
  // single-element union are therefore linear in the current cardinality.
  // Immutable.Map supplies a persistent HAMT while DafnyKey preserves Dafny's
  // structural equality rather than JavaScript object identity.
  class DafnyKey {
    constructor(value) {
      this.value = value;
      this._hash = undefined;
    }
    equals(other) {
      return other instanceof DafnyKey &&
        _dafny.areEqual(this.value, other.value);
    }
    hashCode() {
      if (this._hash === undefined) {
        let hash = 0;
        const text = _dafny.toString(this.value);
        for (let index = 0; index < text.length; index++) {
          hash = ((hash << 5) - hash + text.charCodeAt(index)) | 0;
        }
        this._hash = hash;
      }
      return this._hash;
    }
  }
  $module.Set = class Set {
    constructor(inner) {
      this._inner = inner === undefined ? Immutable.Map() : inner;
    }
    static get Default() {
      return Set.Empty;
    }
    get length() {
      return this._inner.size;
    }
    static get Empty() {
      if (this._empty === undefined) {
        this._empty = new Set();
      }
      return this._empty;
    }
    static fromElements(...elements) {
      const result = new Set();
      for (const element of elements) {
        result.add(element);
      }
      return result;
    }
    static of(...elements) {
      return Set.fromElements(...elements);
    }
    contains(value) {
      return this._inner.has(new DafnyKey(value));
    }
    add(value) {
      const key = new DafnyKey(value);
      const changed = !this._inner.has(key);
      this._inner = this._inner.set(key, value);
      return changed;
    }
    push(...values) {
      for (const value of values) {
        this.add(value);
      }
      return this.length;
    }
    [Symbol.iterator]() {
      return this._inner.valueSeq()[Symbol.iterator]();
    }
    get Elements() {
      return this;
    }
    toString() {
      return "{" + arrayElementsToString(this) + "}";
    }
    equals(other) {
      if (this === other) {
        return true;
      }
      if (!(other instanceof Set) || this.length !== other.length) {
        return false;
      }
      for (const element of this) {
        if (!other.contains(element)) {
          return false;
        }
      }
      return true;
    }
    Union(other) {
      if (this.length === 0) {
        return other;
      }
      if (other.length === 0) {
        return this;
      }
      const larger = this.length >= other.length ? this : other;
      const smaller = larger === this ? other : this;
      const result = new Set(larger._inner);
      for (const element of smaller) {
        result.add(element);
      }
      return result;
    }
    Intersect(other) {
      if (this.length === 0) {
        return this;
      }
      if (other.length === 0) {
        return other;
      }
      const smaller = this.length <= other.length ? this : other;
      const larger = smaller === this ? other : this;
      const result = new Set();
      for (const element of smaller) {
        if (larger.contains(element)) {
          result.add(element);
        }
      }
      return result;
    }
    Difference(other) {
      if (this.length === 0 || other.length === 0) {
        return this;
      }
      let inner = this._inner;
      for (const element of other) {
        inner = inner.remove(new DafnyKey(element));
      }
      return new Set(inner);
    }
    IsDisjointFrom(other) {
      const smaller = this.length <= other.length ? this : other;
      const larger = smaller === this ? other : this;
      for (const element of smaller) {
        if (larger.contains(element)) {
          return false;
        }
      }
      return true;
    }
    IsSubsetOf(other) {
      if (other.length < this.length) {
        return false;
      }
      for (const element of this) {
        if (!other.contains(element)) {
          return false;
        }
      }
      return true;
    }
    IsProperSubsetOf(other) {
      return this.length < other.length && this.IsSubsetOf(other);
    }
    get AllSubsets() {
      return this.AllSubsets_();
    }
    *AllSubsets_() {
      const elements = Array.from(this);
      const count = elements.length;
      const selected = new Array(count).fill(false);
      const current = [];
      while (true) {
        yield Set.fromElements(...current);
        let index = 0;
        for (; index < count && selected[index]; index++) {
          selected[index] = false;
          const removed = current.findIndex(
            value => _dafny.areEqual(value, elements[index]));
          current.splice(removed, 1);
        }
        if (index === count) {
          break;
        }
        selected[index] = true;
        current.push(elements[index]);
      }
    }
  }
